In [1]:
import pandas as pd
import numpy as np

### Load Train and Test Data:

In [2]:
train_y = pd.read_csv(r"/Users/shabnanasser/workplace/git/Capstone_Two/Train-1542865627584.csv")
train_beneficiary = pd.read_csv(r"/Users/shabnanasser/workplace/git/Capstone_Two/Train_Beneficiarydata-1542865627584.csv")
train_inpatient = pd.read_csv(r"/Users/shabnanasser/workplace/git/Capstone_Two/Train_Inpatientdata-1542865627584.csv")
train_outpatient = pd.read_csv(r"/Users/shabnanasser/workplace/git/Capstone_Two/Train_Outpatientdata-1542865627584.csv")

In [3]:
test_y = pd.read_csv(r"/Users/shabnanasser/workplace/git/Capstone_Two/Test-1542969243754.csv")
test_beneficiary = pd.read_csv(r"/Users/shabnanasser/workplace/git/Capstone_Two/Test_Beneficiarydata-1542969243754.csv")
test_inpatient = pd.read_csv(r"/Users/shabnanasser/workplace/git/Capstone_Two/Test_Inpatientdata-1542969243754.csv")
test_outpatient = pd.read_csv(r"/Users/shabnanasser/workplace/git/Capstone_Two/Test_Outpatientdata-1542969243754.csv")

### Shape of data:

In [4]:
print('Shape of train_y data :',train_y.shape)
print('Shape of train_beneficiary data :',train_beneficiary.shape)
print('Shape of train_inpatient data :',train_inpatient.shape)
print('Shape of train_outpatient data :',train_outpatient.shape)

print('Shape of test_y data :',test_y.shape)
print('Shape of test_beneficiary data :',test_beneficiary.shape)
print('Shape of test_inpatient data :',test_inpatient.shape)
print('Shape of test_outpatient data :',test_outpatient.shape)

Shape of train_y data : (5410, 2)
Shape of train_beneficiary data : (138556, 25)
Shape of train_inpatient data : (40474, 30)
Shape of train_outpatient data : (517737, 27)
Shape of test_y data : (1353, 1)
Shape of test_beneficiary data : (63968, 25)
Shape of test_inpatient data : (9551, 30)
Shape of test_outpatient data : (125841, 27)


## train data

In [5]:
train_y.head()

Provider PotentialFraud
0  PRV51001             No
1  PRV51003            Yes
2  PRV51004             No
3  PRV51005            Yes
4  PRV51007             No

In [6]:
train_y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5410 entries, 0 to 5409
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Provider        5410 non-null   object
 1   PotentialFraud  5410 non-null   object
dtypes: object(2)
memory usage: 84.7+ KB


In [7]:
## Lets check whether  providers details are unique or not in train data

print(train_y.Provider.value_counts(sort=True,ascending=False).head(2))  # number of unique providers in train data.Check for duplicates

print('\n Total missing values in train_y :',train_y.isna().sum().sum())

print('\n Total missing values in test_y :',test_y.isna().sum().sum())

Provider
PRV51001    1
PRV55516    1
Name: count, dtype: int64

 Total missing values in train_y : 0

 Total missing values in test_y : 0


## train_beneficiary data

In [8]:
train_beneficiary.head()

BeneID         DOB  DOD  Gender  Race RenalDiseaseIndicator  State  \
0  BENE11001  1943-01-01  NaN       1     1                     0     39   
1  BENE11002  1936-09-01  NaN       2     1                     0     39   
2  BENE11003  1936-08-01  NaN       1     1                     0     52   
3  BENE11004  1922-07-01  NaN       1     1                     0     39   
4  BENE11005  1935-09-01  NaN       1     1                     0     24   

   County  NoOfMonths_PartACov  NoOfMonths_PartBCov  ...  \
0     230                   12                   12  ...   
1     280                   12                   12  ...   
2     590                   12                   12  ...   
3     270                   12                   12  ...   
4     680                   12                   12  ...   

   ChronicCond_Depression  ChronicCond_Diabetes  ChronicCond_IschemicHeart  \
0                       1                     1                          1   
1                       2                     2                          2   
2                       2                     2                          1   
3                       2                     1                          1   
4                       2                     1                          2   

   ChronicCond_Osteoporasis  ChronicCond_rheumatoidarthritis  \
0                         2                                1   
1                         2                                2   
2                         2                                2   
3                         1                                1   
4                         2                                2   

   ChronicCond_stroke  IPAnnualReimbursementAmt  IPAnnualDeductibleAmt  \
0                   1                     36000                   3204   
1                   2                         0                      0   
2                   2                         0                      0   
3                   2                         0                      0   
4                   2                         0                      0   

   OPAnnualReimbursementAmt  OPAnnualDeductibleAmt  
0                        60                     70  
1                        30                     50  
2                        90                     40  
3                      1810                    760  
4                      1790                   1200  

[5 rows x 25 columns]

In [9]:
train_beneficiary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138556 entries, 0 to 138555
Data columns (total 25 columns):
 #   Column                           Non-Null Count   Dtype 
---  ------                           --------------   ----- 
 0   BeneID                           138556 non-null  object
 1   DOB                              138556 non-null  object
 2   DOD                              1421 non-null    object
 3   Gender                           138556 non-null  int64 
 4   Race                             138556 non-null  int64 
 5   RenalDiseaseIndicator            138556 non-null  object
 6   State                            138556 non-null  int64 
 7   County                           138556 non-null  int64 
 8   NoOfMonths_PartACov              138556 non-null  int64 
 9   NoOfMonths_PartBCov              138556 non-null  int64 
 10  ChronicCond_Alzheimer            138556 non-null  int64 
 11  ChronicCond_Heartfailure         138556 non-null  int64 
 12  ChronicCond_Kidn

In [10]:
train_beneficiary.isna().sum()

BeneID                                  0
DOB                                     0
DOD                                137135
Gender                                  0
Race                                    0
RenalDiseaseIndicator                   0
State                                   0
County                                  0
NoOfMonths_PartACov                     0
NoOfMonths_PartBCov                     0
ChronicCond_Alzheimer                   0
ChronicCond_Heartfailure                0
ChronicCond_KidneyDisease               0
ChronicCond_Cancer                      0
ChronicCond_ObstrPulmonary              0
ChronicCond_Depression                  0
ChronicCond_Diabetes                    0
ChronicCond_IschemicHeart               0
ChronicCond_Osteoporasis                0
ChronicCond_rheumatoidarthritis         0
ChronicCond_stroke                      0
IPAnnualReimbursementAmt                0
IPAnnualDeductibleAmt                   0
OPAnnualReimbursementAmt          

In [11]:
#Looking for duplicates in train_beneficiary
train_beneficiary.duplicated().sum()

0

In [12]:
#How many unique beneficiaries we have in our dataset?
train_beneficiary['BeneID'].nunique()

138556

In [13]:
#convert DOB and DOD columns into correct date format

train_beneficiary['DOB'] = pd.to_datetime(train_beneficiary['DOB'], format="%Y-%m-%d")
train_beneficiary['DOD'] = pd.to_datetime(train_beneficiary['DOD'], format="%Y-%m-%d")


In [14]:
#calculate and add AGE column to beneficiary dataframe

train_beneficiary['AGE'] = np.round(((train_beneficiary['DOD'] - train_beneficiary['DOB']).dt.days)/365.0,1)

In [15]:
train_beneficiary.drop(labels=['DOD'],axis=1,inplace=True)

In [16]:
train_beneficiary.head()

BeneID        DOB  Gender  Race RenalDiseaseIndicator  State  County  \
0  BENE11001 1943-01-01       1     1                     0     39     230   
1  BENE11002 1936-09-01       2     1                     0     39     280   
2  BENE11003 1936-08-01       1     1                     0     52     590   
3  BENE11004 1922-07-01       1     1                     0     39     270   
4  BENE11005 1935-09-01       1     1                     0     24     680   

   NoOfMonths_PartACov  NoOfMonths_PartBCov  ChronicCond_Alzheimer  ...  \
0                   12                   12                      1  ...   
1                   12                   12                      2  ...   
2                   12                   12                      1  ...   
3                   12                   12                      1  ...   
4                   12                   12                      2  ...   

   ChronicCond_Diabetes  ChronicCond_IschemicHeart  ChronicCond_Osteoporasis  \
0                     1                          1                         2   
1                     2                          2                         2   
2                     2                          1                         2   
3                     1                          1                         1   
4                     1                          2                         2   

   ChronicCond_rheumatoidarthritis  ChronicCond_stroke  \
0                                1                   1   
1                                2                   2   
2                                2                   2   
3                                1                   2   
4                                2                   2   

   IPAnnualReimbursementAmt  IPAnnualDeductibleAmt  OPAnnualReimbursementAmt  \
0                     36000                   3204                        60   
1                         0                      0                        30   
2                         0                      0                        90   
3                         0                      0                      1810   
4                         0                      0                      1790   

   OPAnnualDeductibleAmt  AGE  
0                     70  NaN  
1                     50  NaN  
2                     40  NaN  
3                    760  NaN  
4                   1200  NaN  

[5 rows x 25 columns]

In [17]:
train_beneficiary['AGE'].describe()

count    1421.000000
mean       74.655735
std        10.669068
min        36.100000
25%        68.300000
50%        75.100000
75%        82.100000
max       100.300000
Name: AGE, dtype: float64

In [18]:
train_beneficiary["AGE"].fillna(0, inplace=True)

In [19]:
#beneficiaries on the basis of 'ChronicCond_Alzheimer'

# 1 means +ve with Chronic ALZH Disease
# 2 means -ve with Chronic ALZH Disease

train_beneficiary.groupby(['ChronicCond_Alzheimer'])['BeneID'].count()

ChronicCond_Alzheimer
1    46026
2    92530
Name: BeneID, dtype: int64

In [20]:
#Total Annual Sum of Max IP Reimbursement for 'ChronicCond_Alzheimer'

train_beneficiary.groupby(['ChronicCond_Alzheimer'])['IPAnnualReimbursementAmt'].sum()

ChronicCond_Alzheimer
1    247213280
2    259949690
Name: IPAnnualReimbursementAmt, dtype: int64

This means if a beneficiary with or without chronic ALZH Disease get admitted then the annual reimbursement paid is approx same.

In [21]:
#Total Annual Sum of Max OP Reimbursement for 'ChronicCond_Alzheimer'

train_beneficiary.groupby(['ChronicCond_Alzheimer'])['OPAnnualReimbursementAmt'].sum()

ChronicCond_Alzheimer
1     74721530
2    105154550
Name: OPAnnualReimbursementAmt, dtype: int64

Annual Sum of OP Reimbursement for non-chronic ALZH Disease beneficiaries is a lot more than others.

In [22]:
#Total Annual Sum of IP deductible amount (amount to pay their own) for 'ChronicCond_Alzheimer'

train_beneficiary.groupby(['ChronicCond_Alzheimer'])['IPAnnualDeductibleAmt'].sum()

ChronicCond_Alzheimer
1    28200854
2    27200388
Name: IPAnnualDeductibleAmt, dtype: int64

Annual Sum of IP deductible amount paid by the patient of non-chronic & chronic ALZH Disease is almost same.

In [23]:
#Total Annual Sum of OP deductible amount for 'ChronicCond_Alzheimer'

train_beneficiary.groupby(['ChronicCond_Alzheimer'])['OPAnnualDeductibleAmt'].sum()

ChronicCond_Alzheimer
1    21502981
2    30832150
Name: OPAnnualDeductibleAmt, dtype: int64

Annual Sum of OP deductible amount paid by the patient of non-chronic is more as compared to chronic ALZH Disease.

In [24]:
#AVG IP Reimbursement Amt for chronic and non-chronic ALZH 

train_beneficiary.groupby(['ChronicCond_Alzheimer'])['IPAnnualReimbursementAmt'].sum() / train_beneficiary.groupby(['ChronicCond_Alzheimer'])['BeneID'].count()

ChronicCond_Alzheimer
1    5371.165863
2    2809.355777
dtype: float64

In [25]:
#AVG OP Reimbursement Amt

train_beneficiary.groupby(['ChronicCond_Alzheimer'])['OPAnnualReimbursementAmt'].sum() / train_beneficiary.groupby(['ChronicCond_Alzheimer'])['BeneID'].count()

ChronicCond_Alzheimer
1    1623.463477
2    1136.437372
dtype: float64

In [26]:
#AVG IP deductible Amt

train_beneficiary.groupby(['ChronicCond_Alzheimer'])['IPAnnualDeductibleAmt'].sum() / train_beneficiary.groupby(['ChronicCond_Alzheimer'])['BeneID'].count()

ChronicCond_Alzheimer
1    612.715726
2    293.962909
dtype: float64

In [27]:
#AVG OP deductible Amt

train_beneficiary.groupby(['ChronicCond_Alzheimer'])['OPAnnualDeductibleAmt'].sum() / train_beneficiary.groupby(['ChronicCond_Alzheimer'])['BeneID'].count()

ChronicCond_Alzheimer
1    467.192044
2    333.212472
dtype: float64

Payer pays huge chunk of expenses specially when a beneficiary gets admitted with or without chronic ALZH disease. To be more precise, in case of +ve it is more than 50%.

In [28]:
#dropping columns NoOfMonths_PartACov and PartBCov, since it is no use

train_beneficiary.drop(["NoOfMonths_PartACov", "NoOfMonths_PartBCov"], axis=1, inplace=True)

In [29]:
train_beneficiary.shape

(138556, 23)

In [30]:
train_beneficiary.head()

BeneID        DOB  Gender  Race RenalDiseaseIndicator  State  County  \
0  BENE11001 1943-01-01       1     1                     0     39     230   
1  BENE11002 1936-09-01       2     1                     0     39     280   
2  BENE11003 1936-08-01       1     1                     0     52     590   
3  BENE11004 1922-07-01       1     1                     0     39     270   
4  BENE11005 1935-09-01       1     1                     0     24     680   

   ChronicCond_Alzheimer  ChronicCond_Heartfailure  ChronicCond_KidneyDisease  \
0                      1                         2                          1   
1                      2                         2                          2   
2                      1                         2                          2   
3                      1                         1                          2   
4                      2                         2                          2   

   ...  ChronicCond_Diabetes  ChronicCond_IschemicHeart  \
0  ...                     1                          1   
1  ...                     2                          2   
2  ...                     2                          1   
3  ...                     1                          1   
4  ...                     1                          2   

   ChronicCond_Osteoporasis  ChronicCond_rheumatoidarthritis  \
0                         2                                1   
1                         2                                2   
2                         2                                2   
3                         1                                1   
4                         2                                2   

   ChronicCond_stroke  IPAnnualReimbursementAmt  IPAnnualDeductibleAmt  \
0                   1                     36000                   3204   
1                   2                         0                      0   
2                   2                         0                      0   
3                   2                         0                      0   
4                   2                         0                      0   

   OPAnnualReimbursementAmt  OPAnnualDeductibleAmt  AGE  
0                        60                     70  0.0  
1                        30                     50  0.0  
2                        90                     40  0.0  
3                      1810                    760  0.0  
4                      1790                   1200  0.0  

[5 rows x 23 columns]

## train_inpatient data

In [31]:
train_inpatient.head()

BeneID   ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0  BENE11001  CLM46614   2009-04-12  2009-04-18  PRV55912   
1  BENE11001  CLM66048   2009-08-31  2009-09-02  PRV55907   
2  BENE11001  CLM68358   2009-09-17  2009-09-20  PRV56046   
3  BENE11011  CLM38412   2009-02-14  2009-02-22  PRV52405   
4  BENE11014  CLM63689   2009-08-13  2009-08-30  PRV56614   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                   26000          PHY390922                NaN   
1                    5000          PHY318495          PHY318495   
2                    5000          PHY372395                NaN   
3                    5000          PHY369659          PHY392961   
4                   10000          PHY379376          PHY398258   

  OtherPhysician AdmissionDt  ... ClmDiagnosisCode_7  ClmDiagnosisCode_8  \
0            NaN  2009-04-12  ...               2724               19889   
1            NaN  2009-08-31  ...                NaN                 NaN   
2      PHY324689  2009-09-17  ...                NaN                 NaN   
3      PHY349768  2009-02-14  ...              25062               40390   
4            NaN  2009-08-13  ...               5119               29620   

  ClmDiagnosisCode_9 ClmDiagnosisCode_10 ClmProcedureCode_1  \
0               5849                 NaN                NaN   
1                NaN                 NaN             7092.0   
2                NaN                 NaN                NaN   
3               4019                 NaN              331.0   
4              20300                 NaN             3893.0   

  ClmProcedureCode_2 ClmProcedureCode_3 ClmProcedureCode_4 ClmProcedureCode_5  \
0                NaN                NaN                NaN                NaN   
1                NaN                NaN                NaN                NaN   
2                NaN                NaN                NaN                NaN   
3                NaN                NaN                NaN                NaN   
4                NaN                NaN                NaN                NaN   

  ClmProcedureCode_6  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  

[5 rows x 30 columns]

In [32]:
train_inpatient.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40474 entries, 0 to 40473
Data columns (total 30 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   BeneID                  40474 non-null  object 
 1   ClaimID                 40474 non-null  object 
 2   ClaimStartDt            40474 non-null  object 
 3   ClaimEndDt              40474 non-null  object 
 4   Provider                40474 non-null  object 
 5   InscClaimAmtReimbursed  40474 non-null  int64  
 6   AttendingPhysician      40362 non-null  object 
 7   OperatingPhysician      23830 non-null  object 
 8   OtherPhysician          4690 non-null   object 
 9   AdmissionDt             40474 non-null  object 
 10  ClmAdmitDiagnosisCode   40474 non-null  object 
 11  DeductibleAmtPaid       39575 non-null  float64
 12  DischargeDt             40474 non-null  object 
 13  DiagnosisGroupCode      40474 non-null  object 
 14  ClmDiagnosisCode_1      40474 non-null

In [33]:
train_inpatient.isna().sum()

BeneID                        0
ClaimID                       0
ClaimStartDt                  0
ClaimEndDt                    0
Provider                      0
InscClaimAmtReimbursed        0
AttendingPhysician          112
OperatingPhysician        16644
OtherPhysician            35784
AdmissionDt                   0
ClmAdmitDiagnosisCode         0
DeductibleAmtPaid           899
DischargeDt                   0
DiagnosisGroupCode            0
ClmDiagnosisCode_1            0
ClmDiagnosisCode_2          226
ClmDiagnosisCode_3          676
ClmDiagnosisCode_4         1534
ClmDiagnosisCode_5         2894
ClmDiagnosisCode_6         4838
ClmDiagnosisCode_7         7258
ClmDiagnosisCode_8         9942
ClmDiagnosisCode_9        13497
ClmDiagnosisCode_10       36547
ClmProcedureCode_1        17326
ClmProcedureCode_2        35020
ClmProcedureCode_3        39509
ClmProcedureCode_4        40358
ClmProcedureCode_5        40465
ClmProcedureCode_6        40474
dtype: int64

### Adding Claim_Clearance_Days

In [34]:
train_inpatient['ClaimStartDt'] = pd.to_datetime(train_inpatient['ClaimStartDt'], format="%Y-%m-%d")
train_inpatient['ClaimEndDt'] = pd.to_datetime(train_inpatient['ClaimEndDt'], format="%Y-%m-%d")

In [35]:
train_inpatient['Claim_Duration'] = (train_inpatient['ClaimEndDt'] - train_inpatient['ClaimStartDt']).dt.days

In [36]:
train_inpatient['Claim_Duration'].describe()

count    40474.000000
mean         5.660251
std          5.628873
min          0.000000
25%          2.000000
50%          4.000000
75%          7.000000
max         36.000000
Name: Claim_Duration, dtype: float64

### Checking the relationship between Amount of Insurance Claim Reimbursed v/s Claim Clearance Days

In [37]:
unq_claim_duration_days = train_inpatient['Claim_Duration'].unique()
unq_claim_duration_days

array([ 6,  2,  3,  8, 17,  5,  4,  7,  1, 11, 35,  9, 15, 12, 24, 23, 14,
        0, 10, 13, 18, 19, 28, 16, 21, 26, 22, 32, 27, 20, 31, 25, 34, 30,
       29, 33, 36])

In [38]:
#for each unique claim duration, how many claims were filed for that specific duration
tot_claims_filed_for_specific_days = pd.DataFrame(train_inpatient.groupby(['Claim_Duration'])['BeneID'].count())
tot_claims_filed_for_specific_days

BeneID
Claim_Duration        
0                  605
1                 4599
2                 6119
3                 6899
4                 4993
5                 3579
6                 2714
7                 2210
8                 1612
9                 1154
10                 916
11                 728
12                 582
13                 518
14                 462
15                 388
16                 303
17                 285
18                 212
19                 174
20                 146
21                 148
22                 121
23                 103
24                  80
25                  67
26                  69
27                  77
28                  71
29                  42
30                  36
31                  24
32                  40
33                  37
34                  28
35                 331
36                   2

In [39]:
tot_insc_amount_for_claim_durations = pd.DataFrame(train_inpatient.groupby(['Claim_Duration'])['InscClaimAmtReimbursed'].sum())
tot_insc_amount_for_claim_durations

InscClaimAmtReimbursed
Claim_Duration                        
0                              3925000
1                             33356000
2                             43987610
3                             56477460
4                             43932690
5                             33962440
6                             29057610
7                             25537200
8                             20637600
9                             15083360
10                            13411460
11                            11063700
12                             9337000
13                             8402600
14                             7707700
15                             6868900
16                             5236800
17                             4915400
18                             4238000
19                             3273900
20                             2205300
21                             2596600
22                             2319700
23                             1815000
24                             1148000
25                             1214000
26                             1171000
27                             1485490
28                             1185000
29                              757000
30                              826000
31                              603000
32                              909000
33                              894000
34                              567500
35                             8157000
36                               31000

In [40]:
claim_clearance_amts = pd.merge(left=tot_claims_filed_for_specific_days, right=tot_insc_amount_for_claim_durations,
                                how='inner',
                                left_on=tot_claims_filed_for_specific_days.index,
                                right_on=tot_insc_amount_for_claim_durations.index)
claim_clearance_amts

key_0  BeneID  InscClaimAmtReimbursed
0       0     605                 3925000
1       1    4599                33356000
2       2    6119                43987610
3       3    6899                56477460
4       4    4993                43932690
5       5    3579                33962440
6       6    2714                29057610
7       7    2210                25537200
8       8    1612                20637600
9       9    1154                15083360
10     10     916                13411460
11     11     728                11063700
12     12     582                 9337000
13     13     518                 8402600
14     14     462                 7707700
15     15     388                 6868900
16     16     303                 5236800
17     17     285                 4915400
18     18     212                 4238000
19     19     174                 3273900
20     20     146                 2205300
21     21     148                 2596600
22     22     121                 2319700
23     23     103                 1815000
24     24      80                 1148000
25     25      67                 1214000
26     26      69                 1171000
27     27      77                 1485490
28     28      71                 1185000
29     29      42                  757000
30     30      36                  826000
31     31      24                  603000
32     32      40                  909000
33     33      37                  894000
34     34      28                  567500
35     35     331                 8157000
36     36       2                   31000

In [41]:
claim_clearance_amts.columns = ['Claim_durations_in_days', 'Total_claims', 'All_Claims_Total_Amount']
claim_clearance_amts.head()

Claim_durations_in_days  Total_claims  All_Claims_Total_Amount
0                        0           605                  3925000
1                        1          4599                 33356000
2                        2          6119                 43987610
3                        3          6899                 56477460
4                        4          4993                 43932690

In [42]:
claim_clearance_amts['Avg_Claim_Insc_Amount'] = np.round(claim_clearance_amts['All_Claims_Total_Amount']/claim_clearance_amts['Total_claims'],2)

In [43]:
claim_clearance_amts.head()

Claim_durations_in_days  Total_claims  All_Claims_Total_Amount  \
0                        0           605                  3925000   
1                        1          4599                 33356000   
2                        2          6119                 43987610   
3                        3          6899                 56477460   
4                        4          4993                 43932690   

   Avg_Claim_Insc_Amount  
0                6487.60  
1                7252.88  
2                7188.69  
3                8186.33  
4                8798.86

### Checking the relationship between Claimed and Admitted Durations with Re-imbursed Amount

In [44]:
train_inpatient['DischargeDt'] = pd.to_datetime(train_inpatient['DischargeDt'], format="%Y-%m-%d")
train_inpatient['AdmissionDt'] = pd.to_datetime(train_inpatient['AdmissionDt'], format="%Y-%m-%d")


In [45]:
train_inpatient['Admitted_Days'] = (train_inpatient['DischargeDt']-train_inpatient['AdmissionDt']).dt.days

In [46]:
claims_with_diff_admitted_and_claimed_dur = train_inpatient[(train_inpatient['Claim_Duration'] != train_inpatient['Admitted_Days'])]
claims_with_diff_admitted_and_claimed_dur

BeneID   ClaimID ClaimStartDt ClaimEndDt  Provider  \
35      BENE11093  CLM82202   2009-01-03 2009-01-12  PRV51842   
2163    BENE18587  CLM37486   2009-02-10 2009-02-11  PRV52020   
3864    BENE24807  CLM82212   2008-12-31 2009-01-11  PRV54111   
4422    BENE26849  CLM70032   2009-10-02 2009-10-07  PRV56647   
5400    BENE30503  CLM67400   2009-09-12 2009-09-14  PRV54228   
5647    BENE31388  CLM40738   2009-03-07 2009-03-08  PRV51490   
5869    BENE32209  CLM58624   2009-07-10 2009-07-17  PRV51249   
7752    BENE39346  CLM82211   2009-02-18 2009-02-26  PRV54065   
7828    BENE39685  CLM47198   2009-04-20 2009-04-22  PRV53705   
7869    BENE39802  CLM73187   2009-11-01 2009-11-05  PRV52100   
9051    BENE44335  CLM76442   2009-11-26 2009-11-27  PRV56259   
9094    BENE44545  CLM82209   2009-01-18 2009-01-27  PRV52843   
9215    BENE45010  CLM32686   2009-01-11 2009-01-23  PRV56778   
9974    BENE47786  CLM82208   2009-05-27 2009-06-04  PRV52828   
11103   BENE51932  CLM44148   2009-03-28 2009-03-29  PRV51282   
11570   BENE53771  CLM75777   2009-11-14 2009-11-17  PRV52152   
12418   BENE56734  CLM59678   2009-07-16 2009-07-17  PRV54772   
12763   BENE58160  CLM41061   2009-03-18 2009-04-08  PRV51501   
13660   BENE61350  CLM67680   2009-09-13 2009-09-20  PRV52279   
13974   BENE62437  CLM65892   2009-08-31 2009-09-02  PRV53959   
16043   BENE69958  CLM64322   2009-08-19 2009-08-24  PRV56583   
17023   BENE73568  CLM31660   2009-01-16 2009-02-03  PRV52821   
17060   BENE73681  CLM82213   2009-02-10 2009-02-21  PRV54354   
17581   BENE75606  CLM31548   2009-01-26 2009-02-02  PRV57166   
18267   BENE78133  CLM82210   2009-01-02 2009-01-05  PRV53461   
18270   BENE78138  CLM72416   2009-10-21 2009-10-22  PRV52469   
21298   BENE88968  CLM82205   2009-02-23 2009-02-28  PRV52045   
22977   BENE95119  CLM64760   2009-08-22 2009-08-24  PRV54258   
24250  BENE100010  CLM47033   2009-04-24 2009-05-01  PRV55916   
25600  BENE104845  CLM36134   2009-02-26 2009-02-27  PRV53706   
26395  BENE107828  CLM82216   2009-01-01 2009-01-22  PRV55018   
27053  BENE110116  CLM82217   2009-02-09 2009-03-17  PRV56268   
27400  BENE111377  CLM36986   2009-02-15 2009-02-20  PRV52019   
27576  BENE112054  CLM45867   2009-04-07 2009-04-16  PRV51119   
28093  BENE114124  CLM52014   2009-05-22 2009-06-06  PRV57399   
28246  BENE114653  CLM44227   2009-04-21 2009-04-30  PRV56449   
28355  BENE115024  CLM75876   2009-11-17 2009-11-19  PRV51497   
28398  BENE115181  CLM42782   2009-03-18 2009-03-19  PRV53710   
28753  BENE116436  CLM82206   2008-12-26 2009-01-04  PRV52178   
29157  BENE117863  CLM39485   2009-03-02 2009-03-09  PRV52266   
29584  BENE119457  CLM82220   2009-02-20 2009-03-03  PRV57547   
29727  BENE119908  CLM82218   2009-01-03 2009-01-29  PRV57191   
31734  BENE127103  CLM60996   2009-08-19 2009-08-28  PRV53226   
32874  BENE131263  CLM82201   2009-01-17 2009-01-25  PRV51021   
34668  BENE138446  CLM82214   2009-01-08 2009-01-12  PRV54955   
35222  BENE140512  CLM76096   2009-11-19 2009-12-06  PRV56268   
39612  BENE156020  CLM82204   2008-12-01 2009-01-06  PRV52019   
39683  BENE156315  CLM46335   2009-04-10 2009-04-11  PRV51037   
39821  BENE156866  CLM82219   2009-01-21 2009-01-28  PRV57409   

       InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
35                      20000          PHY341192          PHY355416   
2163                     3000          PHY359519                NaN   
3864                    13000          PHY360669          PHY408812   
4422                    12000          PHY332134                NaN   
5400                     3000          PHY364179                NaN   
5647                    16000          PHY325912          PHY325912   
5869                     6000          PHY326642          PHY383790   
7752                    19000          PHY337531          PHY341176   
7828                      600          PHY416391          PHY356670   
7869                     3000          PHY366670  

In [47]:
claims_with_diff_admitted_and_claimed_dur['InscClaimAmtReimbursed'].sum()

671400

In [48]:
#checking the rows with more claimed_duration than admitted_days 

claims_with_more_admitted_than_claimed_dur = train_inpatient[(train_inpatient['Claim_Duration'] > train_inpatient['Admitted_Days'])]
print(claims_with_more_admitted_than_claimed_dur.shape[0])


#checking the sum of insurance claim amt where claim duration greater than that of admitted days 

claims_with_diff_admitted_and_claimed_dur[claims_with_diff_admitted_and_claimed_dur['Claim_Duration']  > claims_with_diff_admitted_and_claimed_dur['Admitted_Days']]['InscClaimAmtReimbursed'].sum()


17


274000

In [49]:
#checking the rows with more admitted_days than claimed_duration 

claims_with_more_admitted_than_claimed_dur = train_inpatient[(train_inpatient['Claim_Duration'] < train_inpatient['Admitted_Days'])]
print(claims_with_more_admitted_than_claimed_dur.shape[0])


#checking the sum of insurance claim amt where admitted days  greater than that of Claim_Duration

claims_with_diff_admitted_and_claimed_dur[claims_with_diff_admitted_and_claimed_dur['Claim_Duration']  < claims_with_diff_admitted_and_claimed_dur['Admitted_Days']]['InscClaimAmtReimbursed'].sum()

32


397400

- The above tables tells us that there are 49 claims whose Claimed Duration and Admitted Duration are different. The total re-imbursed amount is around 0.67 Million. 
- 17 claims out of 49 have Claimed Duration greater than the Admitted Duration. And, for these 17 claims the total re-imbursed amount is around 0.27 Million.
- 32 claims out of 49 have Admitted Duration greater than the claimed Duration. And, for these 32 claims the total re-imbursed amount is around 0.39 Million.

### Checking the relationship between DeductibleAmtPaid and Re-imbursed Amount

In [50]:
#number of rows with deductible amt null means no deductibles paid

no_of_claim_with_no_copay = train_inpatient[train_inpatient['DeductibleAmtPaid'].isna()].shape[0]
no_of_claim_with_no_copay

899

In [51]:
#number of rows with deductible amt notnull means deductibles paid

no_of_claim_with_copay = train_inpatient[train_inpatient['DeductibleAmtPaid'].notna()].shape[0]
no_of_claim_with_copay

39575

In [52]:
#reimbursed amt for no co-pay

re_imbursed_amt_for_no_copay = train_inpatient[train_inpatient['DeductibleAmtPaid'].isna()]['InscClaimAmtReimbursed'].sum()
re_imbursed_amt_for_no_copay

10690000

In [53]:
#reimbursed amt with some deductibles

re_imbursed_amt_with_some_copay = train_inpatient[train_inpatient['DeductibleAmtPaid'].notna()]['InscClaimAmtReimbursed'].sum()
re_imbursed_amt_with_some_copay

397607020

In [54]:
tot_sum_of_claims_with_copay = re_imbursed_amt_with_some_copay / (re_imbursed_amt_with_some_copay + re_imbursed_amt_for_no_copay)
tot_sum_of_claims_with_no_copay = re_imbursed_amt_for_no_copay / (re_imbursed_amt_with_some_copay + re_imbursed_amt_for_no_copay)

In [55]:
tot_sum_of_claims_with_copay

0.9738180797890712

In [56]:
tot_sum_of_claims_with_no_copay

0.0261819202109288

The above finfings tells us that there are a 2% of total claims for which there is no co-payment.

And, for these 2% (or 899) of total claims the total re-imbursed amount is 10.6 Million that is 2.6% of the total re-imbursed amount.


In [57]:
train_inpatient['DeductibleAmtPaid'].fillna(value=0.0, inplace=True)

### Checking the relationship of Providers with Total number of claims filed & Re-imbursed Amount

In [58]:
# How many unique providers are there in the inpatient_data?
train_inpatient['Provider'].nunique()

2092

We have 2092 unique number of Providers in the in-patient dataset.

In [59]:
#count of claims filed for each unique provider

provider_tot_claims_filed = pd.DataFrame(train_inpatient.groupby(['Provider'])['ClaimID'].count())
provider_tot_claims_filed

ClaimID
Provider         
PRV51001        5
PRV51003       62
PRV51007        3
PRV51008        2
PRV51011        1
...           ...
PRV57719        4
PRV57728        1
PRV57729       11
PRV57732       11
PRV57757       11

[2092 rows x 1 columns]

In [60]:
#for each unique provider, calculating the total amount of money they have been reimbursed. (for all the claims they filed) 

provider_tot_reimbursed_amt = pd.DataFrame(train_inpatient.groupby(['Provider'])['InscClaimAmtReimbursed'].sum())
provider_tot_reimbursed_amt

InscClaimAmtReimbursed
Provider                        
PRV51001                   97000
PRV51003                  573000
PRV51007                   19000
PRV51008                   25000
PRV51011                    5000
...                          ...
PRV57719                   76000
PRV57728                    6000
PRV57729                  107000
PRV57732                   76000
PRV57757                  109000

[2092 rows x 1 columns]

In [61]:
# merging provider total claims filed with provider total reimbursed amt.

prv_tot_filed_claims_and_tot_reimb_amt = pd.merge(left=provider_tot_claims_filed, right=provider_tot_reimbursed_amt, how='inner',
                                                  left_on=provider_tot_claims_filed.index, right_on=provider_tot_reimbursed_amt.index)
prv_tot_filed_claims_and_tot_reimb_amt


key_0  ClaimID  InscClaimAmtReimbursed
0     PRV51001        5                   97000
1     PRV51003       62                  573000
2     PRV51007        3                   19000
3     PRV51008        2                   25000
4     PRV51011        1                    5000
...        ...      ...                     ...
2087  PRV57719        4                   76000
2088  PRV57728        1                    6000
2089  PRV57729       11                  107000
2090  PRV57732       11                   76000
2091  PRV57757       11                  109000

[2092 rows x 3 columns]

In [62]:
#changing the column names of prv_tot_filed_claims_and_tot_reimb_amt

prv_tot_filed_claims_and_tot_reimb_amt.columns = ['ProviderID', 'Tot_Claims_Filed', 'Tot_Re_Imbursed_Amt']
prv_tot_filed_claims_and_tot_reimb_amt

ProviderID  Tot_Claims_Filed  Tot_Re_Imbursed_Amt
0      PRV51001                 5                97000
1      PRV51003                62               573000
2      PRV51007                 3                19000
3      PRV51008                 2                25000
4      PRV51011                 1                 5000
...         ...               ...                  ...
2087   PRV57719                 4                76000
2088   PRV57728                 1                 6000
2089   PRV57729                11               107000
2090   PRV57732                11                76000
2091   PRV57757                11               109000

[2092 rows x 3 columns]

In [63]:

prv_tot_filed_claims_and_tot_reimb_amt.reset_index(drop=True,inplace=True)

In [64]:
#adding a column for percentage of amt out of total reimbursed amt

prv_tot_filed_claims_and_tot_reimb_amt['Percentage_out_of_tot_reimb_amt'] = round((prv_tot_filed_claims_and_tot_reimb_amt['Tot_Re_Imbursed_Amt'] / 
                                                                                   train_inpatient['InscClaimAmtReimbursed'].sum()) * 100, 3)
prv_tot_filed_claims_and_tot_reimb_amt

ProviderID  Tot_Claims_Filed  Tot_Re_Imbursed_Amt  \
0      PRV51001                 5                97000   
1      PRV51003                62               573000   
2      PRV51007                 3                19000   
3      PRV51008                 2                25000   
4      PRV51011                 1                 5000   
...         ...               ...                  ...   
2087   PRV57719                 4                76000   
2088   PRV57728                 1                 6000   
2089   PRV57729                11               107000   
2090   PRV57732                11                76000   
2091   PRV57757                11               109000   

      Percentage_out_of_tot_reimb_amt  
0                               0.024  
1                               0.140  
2                               0.005  
3                               0.006  
4                               0.001  
...                               ...  
2087                            0.019  
2088                            0.001  
2089                            0.026  
2090                            0.019  
2091                            0.027  

[2092 rows x 4 columns]

In [65]:
#for each unique provider, the maximum amount of money they have been reimbursed for any single claim they filed.

provider_max_reimbursed_amt = pd.DataFrame(train_inpatient.groupby(['Provider'])['InscClaimAmtReimbursed'].max())
provider_max_reimbursed_amt

InscClaimAmtReimbursed
Provider                        
PRV51001                   42000
PRV51003                   57000
PRV51007                   10000
PRV51008                   21000
PRV51011                    5000
...                          ...
PRV57719                   57000
PRV57728                    6000
PRV57729                   24000
PRV57732                   17000
PRV57757                   38000

[2092 rows x 1 columns]

In [66]:
#renaming column names for provider_max_reimbursed_amt dataframe
provider_max_reimbursed_amt.rename(columns={"InscClaimAmtReimbursed": "Max_Re_Imbursed_Amt"}, inplace=True)
provider_max_reimbursed_amt

Max_Re_Imbursed_Amt
Provider                     
PRV51001                42000
PRV51003                57000
PRV51007                10000
PRV51008                21000
PRV51011                 5000
...                       ...
PRV57719                57000
PRV57728                 6000
PRV57729                24000
PRV57732                17000
PRV57757                38000

[2092 rows x 1 columns]

In [67]:
#merging prv_tot_filed_claims_and_tot_reimb_amt with provider_max_reimbursed_amt

prv_tot_filed_claims_tot_max_reimb_amt = pd.merge(left=prv_tot_filed_claims_and_tot_reimb_amt, 
                                                  right=provider_max_reimbursed_amt, how='inner',
                                                  left_on=prv_tot_filed_claims_and_tot_reimb_amt['ProviderID'], 
                                                  right_on=provider_max_reimbursed_amt.index)


In [68]:
#dropping index key_0 column
prv_tot_filed_claims_tot_max_reimb_amt.drop(['key_0'], axis=1, inplace=True)

In [69]:
#calculating difference between total and maximum reimbursed amt
prv_tot_filed_claims_tot_max_reimb_amt['Diff_in_Tot_and_Max'] = prv_tot_filed_claims_tot_max_reimb_amt['Tot_Re_Imbursed_Amt'] - \
prv_tot_filed_claims_tot_max_reimb_amt['Max_Re_Imbursed_Amt']

In [70]:
prv_tot_filed_claims_tot_max_reimb_amt.head()

ProviderID  Tot_Claims_Filed  Tot_Re_Imbursed_Amt  \
0   PRV51001                 5                97000   
1   PRV51003                62               573000   
2   PRV51007                 3                19000   
3   PRV51008                 2                25000   
4   PRV51011                 1                 5000   

   Percentage_out_of_tot_reimb_amt  Max_Re_Imbursed_Amt  Diff_in_Tot_and_Max  
0                            0.024                42000                55000  
1                            0.140                57000               516000  
2                            0.005                10000                 9000  
3                            0.006                21000                 4000  
4                            0.001                 5000                    0

In [71]:
prv_tot_filed_claims_tot_max_reimb_amt.sort_values(by=['Diff_in_Tot_and_Max','Max_Re_Imbursed_Amt','Percentage_out_of_tot_reimb_amt'],
                                                   axis=0, inplace=True)
                                            

In [72]:
prv_tot_filed_claims_tot_max_reimb_amt.head(60)

ProviderID  Tot_Claims_Filed  Tot_Re_Imbursed_Amt  \
290    PRV51630                 1                    0   
805    PRV53329                 1                    0   
860    PRV53511                 1                    0   
1342   PRV55184                 1                    0   
1716   PRV56540                 1                    0   
325    PRV51822                 1                  600   
1583   PRV56038                 1                 1000   
280    PRV51618                 1                 2000   
905    PRV53764                 1                 2000   
1088   PRV54411                 1                 2000   
1243   PRV54932                 1                 2000   
24     PRV51061                 1                 3000   
198    PRV51465                 1                 3000   
377    PRV51958                 1                 3000   
416    PRV52087                 2                 3000   
576    PRV52530                 1                 3000   
602    PRV52608                 1                 3000   
604    PRV52610                 1                 3000   
645    PRV52816                 1                 3000   
720    PRV53050                 2                 3000   
767    PRV53240                 1                 3000   
863    PRV53516                 1                 3000   
881    PRV53626                 1                 3000   
978    PRV54063                 1                 3000   
1037   PRV54262                 1                 3000   
1056   PRV54341                 1                 3000   
1145   PRV54668                 1                 3000   
1246   PRV54935                 3                 3000   
1618   PRV56174                 1                 3000   
1747   PRV56601                 1                 3000   
1813   PRV56762                 1                 3000   
1838   PRV56834                 1                 3000   
2076   PRV57650                 1                 3000   
55     PRV51132                 1                 4000   
162    PRV51398                 1                 4000   
309    PRV51711                 1                 4000   
422    PRV52097                 1                 4000   
697    PRV52970                 1                 4000   
721    PRV53052                 1                 4000   
787    PRV53279                 1                 4000   
842    PRV53440                 1                 4000   
871    PRV53600                 1                 4000   
1010   PRV54164                 1                 4000   
1038   PRV54263                 1                 4000   
1128   PRV54611                 1                 4000   
1363   PRV55233                 1                 4000   
1417   PRV55494                 1                 4000   
1492   PRV55764                 1                 4000   
1629   PRV56227                 1                 4000   
1670   PRV56368                 1                 4000   
1676   PRV56398                 1                 4000   
1736   PRV56576                 1                 4000   
1822   PRV56774                 1                 4000   
1849   PRV56902                 1                 4000   
2060   PRV57594                 1                 4000   
4      PRV51011                 1                 5000   
91     PRV51239                 1                 5000   
250    PRV51572                 1                 5000   
263    PRV51595                 1                 5000   
299    PRV51642                 1                 5000   

      Percentage_out_of_tot_reimb_amt  Max_Re_Imbursed_Amt  \
290                             0.000                    0   
805                             0.000                    0   
860                             0.000                    0   
1342                            0.000                    0   
1716                            0.000                    0   
325                             0.000                  600   
1583                            

The above table showing us the Provider Ids who only filed 1 or 2 claims and got the entire amount re-imbursed.

## triain_out-patients data

In [73]:
train_outpatient.head()

BeneID    ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0  BENE11002  CLM624349   2009-10-11  2009-10-11  PRV56011   
1  BENE11003  CLM189947   2009-02-12  2009-02-12  PRV57610   
2  BENE11003  CLM438021   2009-06-27  2009-06-27  PRV57595   
3  BENE11004  CLM121801   2009-01-06  2009-01-06  PRV56011   
4  BENE11004  CLM150998   2009-01-22  2009-01-22  PRV56011   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                      30          PHY326117                NaN   
1                      80          PHY362868                NaN   
2                      10          PHY328821                NaN   
3                      40          PHY334319                NaN   
4                     200          PHY403831                NaN   

  OtherPhysician ClmDiagnosisCode_1  ... ClmDiagnosisCode_9  \
0            NaN              78943  ...                NaN   
1            NaN               6115  ...                NaN   
2            NaN               2723  ...                NaN   
3            NaN              71988  ...                NaN   
4            NaN              82382  ...                NaN   

  ClmDiagnosisCode_10 ClmProcedureCode_1 ClmProcedureCode_2  \
0                 NaN                NaN                NaN   
1                 NaN                NaN                NaN   
2                 NaN                NaN                NaN   
3                 NaN                NaN                NaN   
4                 NaN                NaN                NaN   

  ClmProcedureCode_3 ClmProcedureCode_4 ClmProcedureCode_5 ClmProcedureCode_6  \
0                NaN                NaN                NaN                NaN   
1                NaN                NaN                NaN                NaN   
2                NaN                NaN                NaN                NaN   
3                NaN                NaN                NaN                NaN   
4                NaN                NaN                NaN                NaN   

  DeductibleAmtPaid  ClmAdmitDiagnosisCode  
0                 0                  56409  
1                 0                  79380  
2                 0                    NaN  
3                 0                    NaN  
4                 0                  71947  

[5 rows x 27 columns]

In [74]:
train_outpatient.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517737 entries, 0 to 517736
Data columns (total 27 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   BeneID                  517737 non-null  object 
 1   ClaimID                 517737 non-null  object 
 2   ClaimStartDt            517737 non-null  object 
 3   ClaimEndDt              517737 non-null  object 
 4   Provider                517737 non-null  object 
 5   InscClaimAmtReimbursed  517737 non-null  int64  
 6   AttendingPhysician      516341 non-null  object 
 7   OperatingPhysician      90617 non-null   object 
 8   OtherPhysician          195046 non-null  object 
 9   ClmDiagnosisCode_1      507284 non-null  object 
 10  ClmDiagnosisCode_2      322357 non-null  object 
 11  ClmDiagnosisCode_3      203257 non-null  object 
 12  ClmDiagnosisCode_4      125596 non-null  object 
 13  ClmDiagnosisCode_5      74344 non-null   object 
 14  ClmDiagnosisCode_6  

In [75]:
train_outpatient.isnull().sum()

BeneID                         0
ClaimID                        0
ClaimStartDt                   0
ClaimEndDt                     0
Provider                       0
InscClaimAmtReimbursed         0
AttendingPhysician          1396
OperatingPhysician        427120
OtherPhysician            322691
ClmDiagnosisCode_1         10453
ClmDiagnosisCode_2        195380
ClmDiagnosisCode_3        314480
ClmDiagnosisCode_4        392141
ClmDiagnosisCode_5        443393
ClmDiagnosisCode_6        468981
ClmDiagnosisCode_7        484776
ClmDiagnosisCode_8        494825
ClmDiagnosisCode_9        502899
ClmDiagnosisCode_10       516654
ClmProcedureCode_1        517575
ClmProcedureCode_2        517701
ClmProcedureCode_3        517733
ClmProcedureCode_4        517735
ClmProcedureCode_5        517737
ClmProcedureCode_6        517737
DeductibleAmtPaid              0
ClmAdmitDiagnosisCode     412312
dtype: int64

### Adding Claim_Clearance_Days

In [76]:
train_outpatient['ClaimStartDt'] = pd.to_datetime(train_outpatient['ClaimStartDt'], format="%Y-%m-%d")
train_outpatient['ClaimEndDt'] = pd.to_datetime(train_outpatient['ClaimEndDt'], format="%Y-%m-%d")

train_outpatient['Claim_Duration'] = (train_outpatient['ClaimEndDt'] - train_outpatient['ClaimStartDt']).dt.days



In [77]:
train_outpatient['Claim_Duration'].describe()

count    517737.000000
mean          1.420532
std           4.707387
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max          23.000000
Name: Claim_Duration, dtype: float64

### Checking the relationship between Amount of Insurance Claim Reimbursed v/s Claim Clearance Days

In [78]:
unq_claim_duration_days = train_outpatient['Claim_Duration'].unique()
unq_claim_duration_days


array([ 0, 20, 12,  7, 14,  1, 15,  3,  6,  2, 13,  5,  9, 18, 10, 19, 16,
        4,  8, 17, 11, 23, 21])

In [79]:
train_outpatient.columns

Index(['BeneID', 'ClaimID', 'ClaimStartDt', 'ClaimEndDt', 'Provider',
       'InscClaimAmtReimbursed', 'AttendingPhysician', 'OperatingPhysician',
       'OtherPhysician', 'ClmDiagnosisCode_1', 'ClmDiagnosisCode_2',
       'ClmDiagnosisCode_3', 'ClmDiagnosisCode_4', 'ClmDiagnosisCode_5',
       'ClmDiagnosisCode_6', 'ClmDiagnosisCode_7', 'ClmDiagnosisCode_8',
       'ClmDiagnosisCode_9', 'ClmDiagnosisCode_10', 'ClmProcedureCode_1',
       'ClmProcedureCode_2', 'ClmProcedureCode_3', 'ClmProcedureCode_4',
       'ClmProcedureCode_5', 'ClmProcedureCode_6', 'DeductibleAmtPaid',
       'ClmAdmitDiagnosisCode', 'Claim_Duration'],
      dtype='object')

In [80]:
#for each unique claim duration, how many claims were filed for that specific duration
tot_claims_filed_for_specific_days = pd.DataFrame(train_outpatient.groupby(['Claim_Duration'])['ClaimID'].count())
tot_claims_filed_for_specific_days

ClaimID
Claim_Duration         
0                453348
1                 11960
2                  4366
3                  2597
4                  2238
5                  1511
6                  1306
7                  2564
8                  1189
9                  1378
10                  982
11                  953
12                  903
13                  947
14                 2735
15                  975
16                 1136
17                  808
18                  800
19                  727
20                24312
21                    1
23                    1

In [81]:
tot_insc_amount_for_claim_durations = pd.DataFrame(train_outpatient.groupby(['Claim_Duration'])['InscClaimAmtReimbursed'].sum())
tot_insc_amount_for_claim_durations

InscClaimAmtReimbursed
Claim_Duration                        
0                            105624910
1                              5381060
2                              2227230
3                              1272800
4                              1270910
5                               976870
6                               571400
7                              1280020
8                               463190
9                               686930
10                              385090
11                              475220
12                              518670
13                              436950
14                             1183090
15                              370550
16                              558680
17                              299910
18                              426380
19                              357230
20                            23365230
21                               11300
23                              102500

In [82]:
claim_clearance_amts = pd.merge(left=tot_claims_filed_for_specific_days, right=tot_insc_amount_for_claim_durations,
                                how='inner',
                                left_on=tot_claims_filed_for_specific_days.index,
                                right_on=tot_insc_amount_for_claim_durations.index)

claim_clearance_amts.columns = ['Claim_durations_in_days', 'Total_claims', 'All_Claims_Total_Amount']
claim_clearance_amts.head()


Claim_durations_in_days  Total_claims  All_Claims_Total_Amount
0                        0        453348                105624910
1                        1         11960                  5381060
2                        2          4366                  2227230
3                        3          2597                  1272800
4                        4          2238                  1270910

In [83]:
claim_clearance_amts['Avg_Claim_Insc_Amount'] = np.round(claim_clearance_amts['All_Claims_Total_Amount']/claim_clearance_amts['Total_claims'],2)


In [84]:
claim_clearance_amts.head()

Claim_durations_in_days  Total_claims  All_Claims_Total_Amount  \
0                        0        453348                105624910   
1                        1         11960                  5381060   
2                        2          4366                  2227230   
3                        3          2597                  1272800   
4                        4          2238                  1270910   

   Avg_Claim_Insc_Amount  
0                 232.99  
1                 449.92  
2                 510.13  
3                 490.10  
4                 567.88

### Checking the relationship between DeductibleAmtPaid and Re-imbursed Amount

In [85]:
no_of_claim_with_no_copay = train_outpatient[train_outpatient['DeductibleAmtPaid']==0].shape[0]
no_of_claim_with_no_copay

496701

In [86]:
no_of_claim_with_copay = train_outpatient[train_outpatient['DeductibleAmtPaid'] != 0].shape[0]
no_of_claim_with_copay


21036

In [87]:
re_imbursed_amt_for_no_copay = train_outpatient[train_outpatient['DeductibleAmtPaid'] == 0]['InscClaimAmtReimbursed'].sum()
re_imbursed_amt_for_no_copay


142391850

In [88]:
re_imbursed_amt_with_some_copay = train_outpatient[train_outpatient['DeductibleAmtPaid'] != 0]['InscClaimAmtReimbursed'].sum()
re_imbursed_amt_with_some_copay

5854270

In [89]:
tot_sum_of_claims_with_copay = re_imbursed_amt_with_some_copay / (re_imbursed_amt_with_some_copay + re_imbursed_amt_for_no_copay)
tot_sum_of_claims_with_copay


0.03949020723105603

In [90]:
tot_sum_of_claims_with_no_copay = re_imbursed_amt_for_no_copay / (re_imbursed_amt_with_some_copay + re_imbursed_amt_for_no_copay)
tot_sum_of_claims_with_no_copay

0.960509792768944

The findings tells us that there are 95% of total claims for which there is no co-payment.

And, for these 95% of total claims the total re-imbursed amount is 142.3 Million.


### Checking the relationship of Providers with Total number of claims filed & Re-imbursed Amount

In [91]:
#How many unique providers are there in the outpatient dataset
train_outpatient['Provider'].nunique()

5012

In [92]:
# #count of claims filed for each unique provider
provider_tot_claims_filed = pd.DataFrame(train_outpatient.groupby(['Provider'])['ClaimID'].count())
provider_tot_claims_filed

ClaimID
Provider         
PRV51001       20
PRV51003       70
PRV51004      149
PRV51005     1165
PRV51007       69
...           ...
PRV57759       28
PRV57760       22
PRV57761       82
PRV57762        1
PRV57763      118

[5012 rows x 1 columns]

In [93]:
#for each unique provider, calculating the total amount of money they have been reimbursed. (for all the claims they filed) 
provider_tot_reimbursed_amt = pd.DataFrame(train_outpatient.groupby(['Provider'])['InscClaimAmtReimbursed'].sum())


In [94]:
# merging provider total claims filed with provider total reimbursed amt.
prv_tot_filed_claims_and_tot_reimb_amt = pd.merge(left=provider_tot_claims_filed, right=provider_tot_reimbursed_amt, how='inner',
                                                  left_on=provider_tot_claims_filed.index, right_on=provider_tot_reimbursed_amt.index)


In [95]:
#changing the column names of prv_tot_filed_claims_and_tot_reimb_amt
prv_tot_filed_claims_and_tot_reimb_amt.columns = ['ProviderID', 'Tot_Claims_Filed', 'Tot_Re_Imbursed_Amt']

#dropping index column
prv_tot_filed_claims_and_tot_reimb_amt.reset_index(drop=True,inplace=True)

#adding percentage column
prv_tot_filed_claims_and_tot_reimb_amt['Percentage_out_of_tot_reimb_amt'] = round((prv_tot_filed_claims_and_tot_reimb_amt['Tot_Re_Imbursed_Amt'] / train_outpatient['InscClaimAmtReimbursed'].sum()) * 100, 3)


In [96]:
#for each unique provider, the maximum amount of money they have been reimbursed for any single claim they filed.

provider_max_reimbursed_amt = pd.DataFrame(train_outpatient.groupby(['Provider'])['InscClaimAmtReimbursed'].max())

provider_max_reimbursed_amt.rename(columns={"InscClaimAmtReimbursed": "Max_Re_Imbursed_Amt"}, inplace=True)


In [97]:
#merging prv_tot_filed_claims_and_tot_reimb_amt and provider_max_reimbursed_amt
prv_tot_filed_claims_tot_max_reimb_amt = pd.merge(left=prv_tot_filed_claims_and_tot_reimb_amt, 
                                                  right=provider_max_reimbursed_amt, how='inner',
                                                  left_on=prv_tot_filed_claims_and_tot_reimb_amt['ProviderID'], 
                                                  right_on=provider_max_reimbursed_amt.index)


In [98]:
prv_tot_filed_claims_tot_max_reimb_amt.drop(['key_0'], axis=1, inplace=True)

#adding the difference column (between total and max reimbursd amt)
prv_tot_filed_claims_tot_max_reimb_amt['Diff_in_Tot_and_Max'] = prv_tot_filed_claims_tot_max_reimb_amt['Tot_Re_Imbursed_Amt'] - \
prv_tot_filed_claims_tot_max_reimb_amt['Max_Re_Imbursed_Amt']

prv_tot_filed_claims_tot_max_reimb_amt.head()

ProviderID  Tot_Claims_Filed  Tot_Re_Imbursed_Amt  \
0   PRV51001                20                 7640   
1   PRV51003                70                32670   
2   PRV51004               149                52170   
3   PRV51005              1165               280910   
4   PRV51007                69                14710   

   Percentage_out_of_tot_reimb_amt  Max_Re_Imbursed_Amt  Diff_in_Tot_and_Max  
0                            0.005                 1500                 6140  
1                            0.022                 3300                29370  
2                            0.035                 3300                48870  
3                            0.189                 4080               276830  
4                            0.010                 3300                11410

In [99]:
#sorting prv_tot_filed_claims_tot_max_reimb_amt

prv_tot_filed_claims_tot_max_reimb_amt.sort_values(by=['Diff_in_Tot_and_Max','Max_Re_Imbursed_Amt','Percentage_out_of_tot_reimb_amt'],
                                                   axis=0, inplace=True)

In [100]:
prv_tot_filed_claims_tot_max_reimb_amt.iloc[200:220]

ProviderID  Tot_Claims_Filed  Tot_Re_Imbursed_Amt  \
4116   PRV56539                 1                  900   
4645   PRV57283                 1                  900   
4387   PRV56906                 1                 1100   
3440   PRV55622                 1                 1500   
521    PRV51689                 1                 1700   
3430   PRV55610                 1                 1700   
3467   PRV55657                 1                 1700   
4701   PRV57357                 1                 1700   
4326   PRV56819                 1                 1900   
5010   PRV57762                 1                 1900   
433    PRV51577                 1                 2100   
1473   PRV52954                 1                 2100   
2480   PRV54295                 1                 3300   
2485   PRV54302                 1                 3300   
3307   PRV55456                 1                 3300   
82     PRV51119                 1                95580   
1980   PRV53646                 3                   20   
739    PRV51987                 2                   30   
4393   PRV56919                 2                   30   
3437   PRV55619                 2                   60   

      Percentage_out_of_tot_reimb_amt  Max_Re_Imbursed_Amt  \
4116                            0.001                  900   
4645                            0.001                  900   
4387                            0.001                 1100   
3440                            0.001                 1500   
521                             0.001                 1700   
3430                            0.001                 1700   
3467                            0.001                 1700   
4701                            0.001                 1700   
4326                            0.001                 1900   
5010                            0.001                 1900   
433                             0.001                 2100   
1473                            0.001                 2100   
2480                            0.002                 3300   
2485                            0.002                 3300   
3307                            0.002                 3300   
82                              0.064                95580   
1980                            0.000                   10   
739                             0.000                   20   
4393                            0.000                   20   
3437                            0.000                   50   

      Diff_in_Tot_and_Max  
4116                    0  
4645                    0  
4387                    0  
3440                    0  
521                     0  
3430                    0  
3467                    0  
4701                    0  
4326                    0  
5010                    0  
433                     0  
1473                    0  
2480                    0  
2485                    0  
3307                    0  
82                      0  
1980                   10  
739                    10  
4393                   10  
3437                   10

The above table showing us the Provider Ids who only filed 1 or 2 claims and got the entire amount re-imbursed.

## Merging the Datasets

In [101]:
# Commom columns in inpatient and outpatient data

common_cols = [col for col in train_inpatient.columns if col in train_outpatient.columns]
len(common_cols)

28

In [102]:
# Merging the IP and OP dataset on the basis of common columns

train_ip_op_data = pd.merge(left=train_inpatient, right=train_outpatient, left_on=common_cols, right_on=common_cols, how="outer")
train_ip_op_data.shape

(558211, 32)

In [103]:
train_ip_op_data

BeneID    ClaimID ClaimStartDt ClaimEndDt  Provider  \
0        BENE11001   CLM46614   2009-04-12 2009-04-18  PRV55912   
1        BENE11001   CLM66048   2009-08-31 2009-09-02  PRV55907   
2        BENE11001   CLM68358   2009-09-17 2009-09-20  PRV56046   
3        BENE11011   CLM38412   2009-02-14 2009-02-22  PRV52405   
4        BENE11014   CLM63689   2009-08-13 2009-08-30  PRV56614   
...            ...        ...          ...        ...       ...   
558206  BENE159198  CLM510792   2009-08-06 2009-08-06  PRV53699   
558207  BENE159198  CLM551294   2009-08-29 2009-08-29  PRV53702   
558208  BENE159198  CLM596444   2009-09-24 2009-09-24  PRV53676   
558209  BENE159198  CLM636992   2009-10-18 2009-10-18  PRV53689   
558210  BENE159198  CLM686139   2009-11-17 2009-11-18  PRV53689   

        InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                        26000          PHY390922                NaN   
1                         5000          PHY318495          PHY318495   
2                         5000          PHY372395                NaN   
3                         5000          PHY369659          PHY392961   
4                        10000          PHY379376          PHY398258   
...                        ...                ...                ...   
558206                     800          PHY364188          PHY364188   
558207                     400          PHY423019          PHY332284   
558208                      60          PHY361063                NaN   
558209                      70          PHY403198                NaN   
558210                      80          PHY419379                NaN   

       OtherPhysician AdmissionDt  ... ClmDiagnosisCode_9  \
0                 NaN  2009-04-12  ...               5849   
1                 NaN  2009-08-31  ...                NaN   
2           PHY324689  2009-09-17  ...                NaN   
3           PHY349768  2009-02-14  ...               4019   
4                 NaN  2009-08-13  ...              20300   
...               ...         ...  ...                ...   
558206      PHY385752         NaT  ...                NaN   
558207            NaN         NaT  ...                NaN   
558208            NaN         NaT  ...                NaN   
558209      PHY419379         NaT  ...                NaN   
558210      PHY419379         NaT  ...                NaN   

        ClmDiagnosisCode_10 ClmProcedureCode_1 ClmProcedureCode_2  \
0                       NaN                NaN                NaN   
1                       NaN             7092.0                NaN   
2                       NaN                NaN                NaN   
3                       NaN              331.0                NaN   
4                       NaN             3893.0                NaN   
...                     ...                ...                ...   
558206                  NaN                NaN                NaN   
558207                  NaN                NaN                NaN   
558208                  NaN                NaN                NaN   
558209                  NaN                NaN                NaN   
558210                  NaN                NaN                NaN   

       ClmProcedureCode_3 ClmProcedureCode_4 ClmProcedureCode_5  \
0                     NaN                NaN                NaN   
1                     NaN                NaN                NaN   
2                     NaN                NaN                NaN   
3                     NaN                NaN                NaN   
4                     NaN                NaN                NaN   
...                   ...                ...                ...   
558206                NaN                NaN                NaN   
558207                NaN                NaN                NaN   
558208                NaN                NaN                NaN   
558209                NaN                NaN                NaN   
558210                NaN                Na

In [104]:
train_ip_op_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 558211 entries, 0 to 558210
Data columns (total 32 columns):
 #   Column                  Non-Null Count   Dtype         
---  ------                  --------------   -----         
 0   BeneID                  558211 non-null  object        
 1   ClaimID                 558211 non-null  object        
 2   ClaimStartDt            558211 non-null  datetime64[ns]
 3   ClaimEndDt              558211 non-null  datetime64[ns]
 4   Provider                558211 non-null  object        
 5   InscClaimAmtReimbursed  558211 non-null  int64         
 6   AttendingPhysician      556703 non-null  object        
 7   OperatingPhysician      114447 non-null  object        
 8   OtherPhysician          199736 non-null  object        
 9   AdmissionDt             40474 non-null   datetime64[ns]
 10  ClmAdmitDiagnosisCode   145899 non-null  object        
 11  DeductibleAmtPaid       558211 non-null  float64       
 12  DischargeDt             40474 

## Merging the IP_OP Dataset with BENE Data

In [105]:
# Joining the IP_OP dataset with the BENE data

train_ip_op_bene_data = pd.merge(left=train_ip_op_data, right=train_beneficiary, left_on='BeneID', right_on='BeneID',how='inner')
train_ip_op_bene_data.shape

(558211, 54)

## Merging the IP_OP_BENE Dataset with train_y provider target label data

In [106]:
# Joining the IP_OP_BENE dataset with the train_y data

train_iobt_data = pd.merge(left=train_ip_op_bene_data, right=train_y, left_on='Provider', right_on='Provider',how='inner')
train_iobt_data.shape

(558211, 55)

## Entire Dataset

In [107]:
train_iobt_data.head()

BeneID    ClaimID ClaimStartDt ClaimEndDt  Provider  \
0  BENE11001   CLM46614   2009-04-12 2009-04-18  PRV55912   
1  BENE16973  CLM565430   2009-09-06 2009-09-06  PRV55912   
2  BENE17521   CLM34721   2009-01-20 2009-02-01  PRV55912   
3  BENE21718   CLM72336   2009-10-17 2009-11-04  PRV55912   
4  BENE22934   CLM73394   2009-10-25 2009-10-29  PRV55912   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                   26000          PHY390922                NaN   
1                      50          PHY365867          PHY327147   
2                   19000          PHY349293          PHY370861   
3                   17000          PHY334706          PHY334706   
4                   13000          PHY390614          PHY323689   

  OtherPhysician AdmissionDt  ... ChronicCond_IschemicHeart  \
0            NaN  2009-04-12  ...                         1   
1            NaN         NaT  ...                         1   
2      PHY363291  2009-01-20  ...                         1   
3            NaN  2009-10-17  ...                         1   
4      PHY363291  2009-10-25  ...                         1   

   ChronicCond_Osteoporasis ChronicCond_rheumatoidarthritis  \
0                         2                               1   
1                         2                               2   
2                         2                               2   
3                         1                               2   
4                         1                               1   

  ChronicCond_stroke IPAnnualReimbursementAmt IPAnnualDeductibleAmt  \
0                  1                    36000                  3204   
1                  2                    24000                  2136   
2                  2                    19000                  1068   
3                  2                    17000                  1068   
4                  2                    27000                  2136   

  OPAnnualReimbursementAmt OPAnnualDeductibleAmt  AGE PotentialFraud  
0                       60                    70  0.0            Yes  
1                      450                   200  0.0            Yes  
2                      100                    20  0.0            Yes  
3                     1050                   540  0.0            Yes  
4                      450                   160  0.0            Yes  

[5 rows x 55 columns]

In [108]:
train_iobt_data.shape

(558211, 55)